### PR TITLE
move determine_number_of_jobs into spack.util.cpus, use it in concretize

### PR DESCRIPTION
**Pull request by haampie**
follow-up to #37608

Now we get something along the lines of `min(#specs, "config:build_jobs from command line" OR min(cpus available, config:build_jobs))`

### CODE REVIEW

**Comment by haampie** [comment]
@spackbot run pipeline

File: lib/spack/spack/util/cpus.py [hunk 1]
```diff
@@ -18,3 +20,36 @@ def cpus_available():
...
+            value is automatically determined.
+    """
+    if not parallel:
+        return 1
+
+    if command_line is None and "command_line" in spack.config.scopes():
+        command_line = spack.config.get("config:build_jobs", scope="command_line")
+
+    if command_line is not None:
+        return command_line
```
Comment (tgamblin) [inline lib/spack/spack/util/cpus.py]:
```suggestion
    if command_line is not None:
        return command_line
    elif "command_line" in spack.config.scopes():
        command_line = spack.config.get("config:build_jobs", scope="command_line")

```

**Review by tgamblin — CHANGES_REQUESTED** [review]

**Comment by haampie** [comment]
@spackbot fix style

File: lib/spack/spack/util/cpus.py [hunk 2]
```diff
@@ -18,3 +20,35 @@ def cpus_available():
         return len(os.sched_getaffinity(0))  # novermin
     except Exception:
         return multiprocessing.cpu_count()
+
+
+def determine_number_of_jobs(
+    parallel=False, command_line=None, config_default=None, max_cpus=None
+):
```
Comment (alalazo) [inline lib/spack/spack/util/cpus.py]: Can we add type-hints to this module and make the arguments here keyword only?

File: lib/spack/spack/util/cpus.py [hunk 3]
```diff
@@ -18,3 +20,35 @@ def cpus_available():
...
+        max_cpus (int or None): maximum number of CPUs available. When None, this
+            value is automatically determined.
+    """
+    if not parallel:
+        return 1
+
+    if command_line is not None:
+        return command_line
+    elif "command_line" in spack.config.scopes():
+        command_line = spack.config.get("config:build_jobs", scope="command_line")
```
Comment (alalazo) [inline lib/spack/spack/util/cpus.py]: I don't get the `elif`. Where is the `command_line` variable being used?

File: var/spack/repos/builtin/packages/gmake/package.py [hunk 4]
```diff
@@ -6,7 +6,6 @@
 import os
 import re
 
-from spack.build_environment import MakeExecutable, determine_number_of_jobs
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gmake/package.py:9]: Where are we getting `MakeExecutable` from?

Comment (alalazo) [reply var/spack/repos/builtin/packages/gmake/package.py:9]: There are other similar changes below.

**Review by alalazo — CHANGES_REQUESTED** [review]

File: lib/spack/spack/util/cpus.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/util/cpus.py]: I think this suggestion introduced a bug, see https://github.com/spack/spack/pull/37620#discussion_r1316896496

File: var/spack/repos/builtin/packages/gmake/package.py (hunk 4 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/gmake/package.py:9]: lib/spack/spack/package.py

File: lib/spack/spack/util/cpus.py (hunk 3 above)
Comment (haampie) [reply lib/spack/spack/util/cpus.py]: It's for tests

Comment (haampie) [reply lib/spack/spack/util/cpus.py]: Ah, I see what you mean :laughing:

File: var/spack/repos/builtin/packages/gmake/package.py (hunk 4 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/gmake/package.py:9]: I see, it was _already_ there - so not visible in the diff :slightly_smiling_face:

File: lib/spack/spack/util/cpus.py [hunk 5]
```diff
@@ -18,3 +21,36 @@ def cpus_available():
...
     except Exception:
         return multiprocessing.cpu_count()
+
+
+def determine_number_of_jobs(
+    *,
+    parallel: bool = False,
+    max_cpus: int = cpus_available(),
+    config: Optional["spack.config.Configuration"] = None,
+):
```
Comment (alalazo) [inline lib/spack/spack/util/cpus.py]:
```suggestion
) -> int:
```
?

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
@spackbot run pipeline

**Pull request merged**